### PR TITLE
[NEW] Disable user typing status (#12423)

### LINKED ISSUE #12423: Disable my user typing status
I don't want send this message, that I typing a text. It can be disabled in skype. Add this check feature please.
Also for me it will be enough to edit some client's file if it will help to disable this.

image

### PR DESCRIPTION
**Pull request by computerManiac**
Closes #12423 

Rocket.Chat by default shows the **_user is typing_** status whenever a user starts typing as seen below.

screenshot_2018-10-26 dev-server

This fix allows users to disable it. To disable it, go to _packages/rocketchat-ui-message/client/messageBox.js_ and on line 7 we have the following constant defined:
```javascript
const disableUserTyping = false;
```
Changing the value to **true** will disable user typing status.

### CODE REVIEW

**Comment by computerManiac** [comment]
> However, I do not think this should be something that can be turned off. Realizing that someone else is typing is very vital to several conversation flows and thus not having it would actually be a hindrance instead of a help.

I think we should let users have the ability to disable this, but if an admin feels this feature is vital for conversation then he can deny users from disabling it on the server.

**Comment by graywolf336** [comment]
I'm going to stand behind my original statement and say that I'm not sure we should even consider this feature. Otherwise, it can inadvertently lead to miscommunications.

**Comment by computerManiac** [comment]
Should I close the PR ?

**Comment by graywolf336** [comment]
No sir, we will await responses from the rest of the Rocket.Chat team 🙂

**Comment by reetp** [comment]
Personally I can't see the harm in this if the ultimate arbiter is the admin, and someone has already written the code.

We are all different and have differing requirements.

**Comment by computerManiac** [comment]
I have added the option under _Message_ section of Admin Settings to Allow Users to disable typing status. The users can disable it under account settings, if allowed by the admin otherwise the _Disable User Typing_ remains false. There is however a slight problem which I am unable to fix. I have added the corresponding entries in _en.i18n.json_ and the text displayed correctly for the first run but for subsequent runs, the text didn't display correctly. Disabling user typing status unsubscribes the user from the typing stream, so he/she won't be able to see if other users are typing.

**Comment by vynmera** [comment]
why not an administration key? in Message or w/e?

**Comment by reetp** [comment]
Any more comments on this from the devs?

File: packages/rocketchat-ui/client/lib/msgTyping.js [hunk 1]
```diff
@@ -24,6 +31,10 @@ export const MsgTyping = (function() {
 	};
 
 	const addStream = function(room) {
+		if (disableUserTyping === true) {
+			RocketChat.Notifications.unRoom(room, 'typing', function() { return; });
```
Comment (tassoevan) [inline packages/rocketchat-ui/client/lib/msgTyping.js]: The third argument must be the original event listener callback attached to the event, e.g.,

```js
// The event listener callback (cb) is attached to the event
const cb = () => { ... };
RocketChat.Notifications.onRoom(room, 'typing', cb);

...

// The same event listener callback is detached
RocketChat.Notifications.unRoom(room, 'typing', cb);
```

File: packages/rocketchat-ui/client/lib/msgTyping.js [hunk 2]
```diff
@@ -12,6 +12,13 @@ export const MsgTyping = (function() {
 	const selfTyping = new ReactiveVar(false);
 	const usersTyping = {};
 	const dep = new Tracker.Dependency;
+	let disableUserTyping = undefined;
+	Tracker.autorun(function() {
+		if (!Meteor.userId()) {
+			return;
+		}
+		disableUserTyping = RocketChat.getUserPreference(Meteor.userId(), 'disableUserTyping');
```
Comment (tassoevan) [inline packages/rocketchat-ui/client/lib/msgTyping.js]: If `Allow_DisableUserTyping` changes from `true` to `false`, a user can still have the `disableUserTyping` preference set to `true`.

```suggestion
		disableUserTyping = RocketChat.settings.get('Allow_DisableUserTyping') === true 
 ? RocketChat.getUserPreference(Meteor.userId(), 'disableUserTyping') : false;
```

**Comment by rodrigok** [comment]
Closing this one in favor of https://github.com/RocketChat/Rocket.Chat/pull/13232

Thanks

**Pull request closed without merging**